src/weight_windows.cpp

```diff
@@ -0,0 +1,314 @@
+    biasing_energy = get_node_array<double>(node_user_defined_biasing, "biasing_energy");
+    biasing_energy.insert(biasing_energy.begin(), 0.);
+  } else {
+    fatal_error("Must provide energy groups for biasing.");
+  }
+
+  // origin probability for each energy group
+  if (check_for_node(node_user_defined_biasing, "origin_probability")) {
+    auto value = get_node_xarray<double>(node_user_defined_biasing, "origin_probability");
+    if (value.size() != biasing_energy.size() - 1) fatal_error("Origin probabilities and biasing energies must have the same number of values.");
+    origin_probability.push_back(0.);
+    cumulative_probability.push_back(0.);
+    for (int i = 0; i < value.size(); i++) {
+      origin_probability.push_back(value.at(i));
+      cumulative_probability.push_back(0.);
+    }
+    // normalization
+    double total_probability = std::accumulate(origin_probability.begin(), origin_probability.end(), 0.0);
+    for (int i = 1; i < origin_probability.size(); i++) {   
+      origin_probability.at(i) = origin_probability.at(i)/total_probability;
+      cumulative_probability.at(i) = cumulative_probability.at(i-1) + origin_probability.at(i);
+    }
+  } else {
+    fatal_error("Must provide origin_probability for each group.");
+  }
+      
+  // biasing weight for each energy group
+  if (check_for_node(node_user_defined_biasing, "biasing")) {
+    auto value = get_node_xarray<double>(node_user_defined_biasing, "biasing");
+    if (value.size() != biasing_energy.size() - 1) {
+      fatal_error("Biasing and biasing energies must have the same number of values.");
+    }
+    biasing.push_back(0);
+    cumulative_biasing.push_back(0);
+    for (double v : value) {
+      biasing.push_back(v);
+      cumulative_biasing.push_back(0);
+    }
+    // normalization
+    double total_probability = std::accumulate(biasing.begin(), biasing.end(), 0.0);
+    for (int i = 1; i < biasing.size(); i++) {
+      biasing[i] = biasing[i]/total_probability;
+      cumulative_biasing[i] = cumulative_biasing[i-1] + biasing[i];
+    }
+  } else {
+    fatal_error("Must provide biasing for each energy group.");
+  }
+}
+  
+//! source weight biasing in energy
+void WeightWindow::weight_biasing(SourceSite& site, uint64_t* seed) 
+{
+  int i = 0;
+  double r = prn(seed);
+  for (i = 0; i < cumulative_biasing.size() - 1; i++) 
+    if (cumulative_biasing[i] <= r && r < cumulative_biasing[i+1])  break;
+  site.E = biasing_energy[i] + ( biasing_energy[i+1] - biasing_energy[i] ) * prn(seed);
+  site.wgt *= origin_probability[i+1] / biasing[i+1];
+}
+  
+//! Get weight windows parameters given particle - essentially
+// given a location tell the particle the right bounds
+ParticleWeightParams WeightWindow::get_params(Particle& p) const
+{	
+  // Particle's position and energy
+  Position pos  = p.r();
+  double E = p.E();
+
+  // get the settings for the weight window
+  ParticleType type = p.type();
+
+  // 
+  WWParams ww_settings;
```

Comment:
This is a potentially big class since it stores `std::vector`s. Making a copy of it here is perhaps a source of the slowdown relative to normal OpenMC tracking you're seeing.